unittests/AST/DiagnosticConsumerTests.cpp

```diff
@@ -0,0 +1,361 @@
+                               "remark", {}, DiagnosticInfo());
+  topConsumer.finishProcessing();
+}
+
+TEST(FileSpecificDiagnosticConsumer, SameFileNotes) {
+  SourceManager sourceMgr;
+  //                                             01234
+  unsigned bufferA = sourceMgr.addMemBufferCopy("abcde", "A");
+  unsigned bufferB = sourceMgr.addMemBufferCopy("vwxyz", "B");
+
+  SourceLoc frontOfA = sourceMgr.getLocForOffset(bufferA, 0);
+  SourceLoc middleOfA = sourceMgr.getLocForOffset(bufferA, 2);
+  SourceLoc backOfA = sourceMgr.getLocForOffset(bufferA, 4);
+
+  SourceLoc frontOfB = sourceMgr.getLocForOffset(bufferB, 0);
+  SourceLoc middleOfB = sourceMgr.getLocForOffset(bufferB, 2);
+  SourceLoc backOfB = sourceMgr.getLocForOffset(bufferB, 4);
+
+  ExpectedDiagnostic expectedA[] = {
+    {frontOfA, "error"},
+    {middleOfA, "note"},
+    {backOfA, "note"},
+    {frontOfB, "error"},
+    {middleOfB, "note"},
+    {backOfB, "note"},
+    {frontOfA, "error"},
+    {middleOfA, "note"},
+    {backOfA, "note"},
+  };
+  ExpectedDiagnostic expectedB[] = {
+    {frontOfB, "error"},
+    {middleOfB, "note"},
+    {backOfB, "note"},
+  };
+
+  auto consumerA = llvm::make_unique<ExpectationDiagnosticConsumer>(
+      nullptr, expectedA);
+  auto consumerB = llvm::make_unique<ExpectationDiagnosticConsumer>(
+      consumerA.get(), expectedB);
+
+  FileSpecificDiagnosticConsumer::ConsumerPair consumers[] = {
+    {sourceMgr.getRangeForBuffer(bufferA), std::move(consumerA)},
+    {CharSourceRange(), std::move(consumerB)}};
+
+  FileSpecificDiagnosticConsumer topConsumer(consumers);
+  topConsumer.handleDiagnostic(sourceMgr, frontOfA, DiagnosticKind::Error,
+                               "error", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, middleOfA, DiagnosticKind::Note,
+                               "note", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, backOfA, DiagnosticKind::Note,
+                               "note", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, frontOfB, DiagnosticKind::Error,
+                               "error", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, middleOfB, DiagnosticKind::Note,
+                               "note", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, backOfB, DiagnosticKind::Note,
+                               "note", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, frontOfA, DiagnosticKind::Error,
+                               "error", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, middleOfA, DiagnosticKind::Note,
+                               "note", {}, DiagnosticInfo());
+  topConsumer.handleDiagnostic(sourceMgr, backOfA, DiagnosticKind::Note,
+                               "note", {}, DiagnosticInfo());
+  topConsumer.finishProcessing();
+}
+
+TEST(FileSpecificDiagnosticConsumer, CrossFileNotes) {
```

Comment:
I find these test routines a bit daunting. Would it make sense to add comments and/or break them up a bit? I'm thinking of a comment or subroutine name like: "Ensure that a Note ends up in the same consumer as the most recent non-note." Or some such, so a reader can gather the intention without delving into the code.

Another approach: list the ways you can imagine it failing, and test each one. That's probably what this is, but it's hard to tell at a glance. 

At a glance, there might be a factoring that could make these test routines a lot more concise. Just a thought...

Reply:
I admit to testing multiple things in each function, since the setup is so clumsy. But ownership and having all the expectations up front makes it hard to make the setup *not* clumsy. Maybe making the diagnostic *emission* less noisy would help, so you could compare the expected per-file output with what's going to be emitted?

Reply:
No, it turns out I can come up with reasonable names for all these tests.